Add %v for formatting vectors

MarcusElg opened the pull request:
Adds %v for formatting vectors. It works similarly to %f but is applied to each component of the vector. Supports all the formatting options that %f does such as amount of decimals and padding. A few examples:

![image](https://user-images.githubusercontent.com/19669673/182034578-b29f032d-1305-428e-a9a1-1e19ef312377.png)
Prints:
![image](https://user-images.githubusercontent.com/19669673/182034587-6a8b1a27-5e50-4cba-aa10-7b13f61f6272.png)

It works for Vector2/Vector3/Vector4/Vector2i/Vector3i/Vector4i. It treats the integer vectors as if they were regular vectors, just like %f treats integers. If integer vectors and floating point vectors should be treated different then it would make sense to have it as a seperate formatting character.

Implements https://github.com/godotengine/godot-proposals/issues/4972

MarcusElg committed:
Add %v for formatting vectors

xix-xeaon commented:
Doesn't %f work with integers too? There's no reason to exclude integer vectors.

MarcusElg commented:
> Doesn't %f work with integers too? There's no reason to exclude integer vectors.

If that works it's likely due to automatic type conversions as the code reads the value as a double. 

The integer vectors don't automatically convert to normal vectors as far as I know. I could make %v work with integer vectors as well but it would be intresting to hear what other people think of this.

MarcusElg commented:
This should still make it into 4.0 right?

MarcusElg commented:
@akien-mga sorry for ping but can this get some sort of review? As the reactions show people are excited for this feature so it would be nice if it could be merged before 4.0 goes into beta and feature freeze.

tinmanjuggernaut commented:
Woo-hoo, thank you! Manually formatting vectors for print has been super annoying. I need it! 

Yes, please support integer vectors as well. The usefulness comes in saving us from typing out every component. Alignment is also valuable. Any decimal points paramers can just be ignored in case the same print code is used for any vector type. 

Does it properly handle Nan and INF vectors?

MarcusElg commented:
> Woo-hoo, thank you! Manually formatting vectors for print has been super annoying. I need it!
> 
> Yes, please support integer vectors as well. The usefulness comes in saving us from typing out every component. Alignment is also valuable. Any decimal points paramers can just be ignored in case the same print code is used for any vector type.
> 
> Does it properly handle Nan and INF vectors? Does it print arrays of vectors w/ precision? (less important IMO).

Thanks for your support :D 

I have updated it to work for integer vectors as well, just like %f works with integers. However if they should be treated differently (without decimals), like %d compared to %f then there should probably be a different formatting char for that. And in that case should there also be a vector version for octal and hexadecimal? It quickly becomes much more complex so lets keep it simple for now. 

As for arrays of vectors that is also a different use case, which could be implemented in the future. All other formatting only works on a single typ tho, so it would be a bit confusing for %v to work both on vectors and arrays of vectors.

As for vectors with NaN and INF in them, they aren't really formatted correctly, but that is a problem with %f and should be tracked in its own bug report.

So overall I've tried to keep it simple, by working similarly to %f this can hopefully still make it's way into 4.0.

aaronfranke commented on core/string/ustring.cpp:
```diff
@@ -4664,6 +4664,75 @@ String String::sprintf(const Array &values, bool *error) const {
+
+					Array numberValues = Array();
+
+					// Get values from the correct type of vector
+					switch (values[value_index].get_type()) {
+						case Variant::VECTOR2:
+						case Variant::VECTOR2I: {
+							Vector2 value = values[value_index];
+							numberValues.push_back(value.x);
+							numberValues.push_back(value.y);
+
+							break;
+						}
+						case Variant::VECTOR3:
+						case Variant::VECTOR3I: {
+							Vector3 value = values[value_index];
+							numberValues.push_back(value.x);
+							numberValues.push_back(value.y);
+							numberValues.push_back(value.z);
+
+							break;
+						}
+						case Variant::VECTOR4:
+						case Variant::VECTOR4I: {
+							Vector4 value = values[value_index];
+							numberValues.push_back(value.x);
+							numberValues.push_back(value.y);
+							numberValues.push_back(value.z);
+							numberValues.push_back(value.w);
+
+							break;
+						}
+						default: {
+							return "a vector is required";
+						}
+					}
+
+					String str = "(";
+					for (int i = 0; i < numberValues.size(); i++) {
+						double val = numberValues[i];
```
Creating an `Array` seems overkill, we can do it in a simpler way:

```suggestion
					int count;
					// Get values from the correct type of vector
					switch (values[value_index].get_type()) {
						case Variant::VECTOR2:
						case Variant::VECTOR2I: {
							count = 2;
						} break;
						case Variant::VECTOR3:
						case Variant::VECTOR3I: {
							count = 3;
						} break;
						case Variant::VECTOR4:
						case Variant::VECTOR4I: {
							count = 4;
						} break;
						default: {
							return "%v requires a vector type (Vector2/3/4/2i/3i/4i)";
						}
					}
					Vector4 vec = values[value_index];
					String str = "(";
					for (int i = 0; i < count; i++) {
						real_t val = vec[i];
```

MarcusElg replied:
That's much smarter! Thanks a lot! I didn't know that you could parse any vector type as a Vector4.

aaronfranke commented on core/string/ustring.cpp:
```diff
@@ -4664,6 +4664,75 @@ String String::sprintf(const Array &values, bool *error) const {
+					for (int i = 0; i < numberValues.size(); i++) {
+						double val = numberValues[i];
+						// Pad decimals out.
+						String numberStr = String::num(ABS(val), min_decimals).pad_decimals(min_decimals);
```
Use snake_case:

```suggestion
						String number_str = String::num(ABS(val), min_decimals).pad_decimals(min_decimals);
```

MarcusElg replied:
Changed

akien-mga commented on core/string/ustring.cpp:
```diff
@@ -4770,8 +4823,9 @@ String String::sprintf(const Array &values, bool *error) const {
 
-					if (!values[value_index].is_num()) {
-						return "* wants number";
+					Variant::Type valueType = values[value_index].get_type();
```
Use snake case for identifiers.

MarcusElg replied:
Fixed

akien-mga commented on core/string/ustring.cpp:
```diff
@@ -4770,8 +4823,9 @@ String String::sprintf(const Array &values, bool *error) const {
-					if (!values[value_index].is_num()) {
-						return "* wants number";
+					Variant::Type valueType = values[value_index].get_type();
+					if (!values[value_index].is_num() && valueType != Variant::VECTOR2 && valueType != Variant::VECTOR3 && valueType != Variant::VECTOR4 && valueType != Variant::VECTOR2I && valueType != Variant::VECTOR3I && valueType != Variant::VECTOR4I) {
```
```suggestion
					if (!values[value_index].is_num()
							&& valueType != Variant::VECTOR2 && valueType != Variant::VECTOR2I
							&& valueType != Variant::VECTOR3 && valueType != Variant::VECTOR3I
							&& valueType != Variant::VECTOR4 && valueType != Variant::VECTOR4I) {
```
a bit more readable.

MarcusElg replied:
Sure

akien-mga reviewed (commented):
Looks OK, but you should also added test cases for this to validate it.

akien-mga commented on core/string/ustring.cpp:
```diff
@@ -4770,8 +4835,13 @@ String String::sprintf(const Array &values, bool *error) const {
-						return "* wants number";
+					Variant::Type value_type = values[value_index].get_type();
+					if (!values[value_index].is_num()
+							&& value_type != Variant::VECTOR2
+							&& value_type != Variant::VECTOR3
+							&& value_type != Variant::VECTOR4 && value_type != Variant::VECTOR2I
+							&& value_type != Variant::VECTOR3I && value_type != Variant::VECTOR4I) {
```
That looks inconsistent.

I suggested:
```
&& 2 && 2I
&& 3 && 3I
&& 4 && 4I
```
It can also be this if you prefer, though it's less common in the codebase to separate float and int variants in such groups:
```
&& 2 && 3 && 4
&& 2I && 3I && 4I
```

MarcusElg replied:
Fixed. I was a bit too quick there. Sorry

MarcusElg commented:
> Looks OK, but you should also added test cases for this to validate it.

I added a bunch of testcases which also made me realised that it didn't work correctly for vectors with negative values, so I fixed that as well.

akien-mga reviewed (approved):
Seems good.

akien-mga commented:
Thanks!

just-like-that commented:
@MarcusElg : You are awesome. Many thanks!

just-like-that commented:
> As for vectors with NaN and INF in them, they aren't really formatted correctly, but that is also a problem with %f as well and should be tracked in its own bug report. They are printed as nan.000000

[GDScript: print formatted of INF variable leads to weird output](https://github.com/godotengine/godot/issues/62908)

just-like-that commented:
For support of arrays see:
[](https://github.com/godotengine/godot-proposals/issues/5246)
